### PR TITLE
Feature Visualization: Magnitude Constrained optimisation (MaCo)

### PR DESCRIPTION
**Pull request by fel-thomas**
# Unlocking Feature Visualization for Deep Networks with Magnitude-Constrained Optimization

In this PR, I propose the introduction of a recent method for scaling up feature visualization on state-of-the-art deep models: **MaCo**.
This method is described in the following arXiv paper: https://arxiv.org/pdf/2306.06805.pdf. It involves fixing the amplitude in the Fourier spectrum and only optimizing the phase during the optimization of a neuron/channel/layer.

There are no major technical challenges, and it consists of a series of simple commits.
The PR is accompanied by the outstanding @thib-s, who is the author of many tricks that make MaCo work.

# Refactoring 

Commit d00b127c3a558e2c6659ddcb3d57d65ed8d40d31: Initially, the function `_clip_percentile` was restricted to heatmaps. However, we have found that it is also necessary to eliminate extreme values from our optimized images. I propose removing this restriction and allowing clip percentile on tensors as well. As it is a private, non-exposed function, I believe it should not cause any issues. What do you think @AntoninPoche ?

# Core

Commit 1030669eb874aa031248d2766963a8924d14cd10: I am introducing the necessary preconditioning for MaCo, as well as initializing the buffer. @lucashervier, I request your review to see how I loaded an external file. I have done this in several repositories before, such as https://github.com/serre-lab/Harmonization. Keras handles caching for us efficiently, and I don't think we need to create our own method. Please let me know your thoughts.

Commit d36f7a1ea45f4f2d6d9e5b3a4e1ddbce4bc1416a: This commit includes the optimization loop for MaCo. @thib-s, I have tested the notebook alongside the default models, and it passes all tests. Please review and provide your feedback.

# Docs and utils

Commit 0f0dd483ccb9ecf9a5067e6309523c58bea7e720: This commit introduces the plot function that manages the blending alpha. @thib-s, I would appreciate your review on this plot function as well. @AntoninPoche, you have contributed significantly to the plot module. If you have any insights to share, please don't hesitate.

Commit 75a05da523745415804be9cddce52120ebcf68d0: Finally, this commit adds the documentation page and updates the readme to include the notebook.

# Miscellaneous

I finaly had the time to make an update to the Numkdocs parser https://github.com/fel-thomas/numkdoc and also updated it on PyPI. This update now allows us **to have documentation not only on classes but also on functions** (with the same syntax) 🎉!

---

I will wait for your (@thib-s, @AntoninPoche and @lucashervier) validation before merging. ;) 
PS: MaCo PR means that the CRAFT one is coming soon. ^^

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -97,6 +97,7 @@ We propose some Hands-on tutorials to get familiar with the library and its api:
     </a>
 </p>
 
+- [**Modern Feature Visualization with MaCo**: Getting started](https://colab.research.google.com/drive/1l0kag1o-qMY4NCbWuAwnuzkzd9sf92ic) <sub> [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/1l0kag1o-qMY4NCbWuAwnuzkzd9sf92ic) </sub>
```
Comment (lucashervier) [inline README.md:100]: Even though it is probably better, it would be logical that it comes after the Getting Started

File: docs/index.md [hunk 2]
```diff
@@ -98,6 +98,7 @@ We propose some Hands-on tutorials to get familiar with the library and its api:
     </a>
 </p>
 
+- [**Modern Feature Visualization with MaCo**: Getting started](https://colab.research.google.com/drive/1l0kag1o-qMY4NCbWuAwnuzkzd9sf92ic) <sub> [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/1l0kag1o-qMY4NCbWuAwnuzkzd9sf92ic) </sub>
```
Comment (lucashervier) [inline docs/index.md:101]: Idem than for the README

**Review by lucashervier — COMMENTED** [review]
LGTM :)

I am just questionning the order of appearance of the Feat Viz notebook that I would like to change.

**Review by Agustin-Picard — APPROVED** [review]
LGTM, great work @fel-thomas as usual ;)

**Comment by fel-thomas** [comment]
Thanks everyone

> LGTM :)
> 
> I am just questionning the order of appearance of the Feat Viz notebook that I would like to change.

Done :)

**Pull request merged**